sling/nlp/kb/phrase-table-builder.cc

```diff
@@ -0,0 +1,466 @@
+    repository.AddBlock("normalization", norm.data(), norm.size());
+
+    // Write entity map.
+    LOG(INFO) << "Build entity map";
+    File *entity_index_block = repository.AddBlock("EntityIndex");
+    File *entity_item_block = repository.AddBlock("EntityItems");
+    uint32 offset = 0;
+    for (Entity &entity : entity_table_) {
+      // Write entity index entry.
+      entity_index_block->WriteOrDie(&offset, sizeof(uint32));
+
+      // Write count and id to entity entry.
+      CHECK_LT(entity.id.size(), 256);
+      uint8 idlen = entity.id.size();
+      entity_item_block->WriteOrDie(&entity.count, sizeof(uint32));
+      entity_item_block->WriteOrDie(&idlen, sizeof(uint8));
+      entity_item_block->WriteOrDie(entity.id.data(), idlen);
+
+      // Compute offset of next entry.
+      offset += sizeof(uint32) + sizeof(uint8) + idlen;
+    }
+
+    // Write phrase map.
+    LOG(INFO) << "Build phrase map";
+    int num_phrases = phrase_table_.size();
+    int num_buckets = (num_phrases + 32) / 32;
+    std::vector<RepositoryMapItem *> items;
+    items.reserve(num_phrases);
+    for (auto &it : phrase_table_) {
+      Phrase *phrase = it.second;
+      items.push_back(phrase);
+
+      // Sort entities in decreasing order.
+      std::sort(phrase->entities.begin(), phrase->entities.end(),
+          [](const EntityPhrase &a, const EntityPhrase &b) {
+            return a.count() > b.count();
+          });
+    }
+    repository.WriteMap("Phrase", &items, num_buckets);
+
+    // Write repository to file.
+    const string &filename = task->GetOutput("repository")->resource()->name();
+    CHECK(!filename.empty());
+    LOG(INFO) << "Write phrase repository to " << filename;
+    repository.Write(filename);
+    LOG(INFO) << "Repository done";
+
+    // Clear collected data.
+    for (auto &it : phrase_table_) delete it.second;
+    phrase_table_.clear();
+    entity_table_.clear();
+    entity_mapping_.clear();
+  }
+
+ private:
+  // Entity with id and frequency.
+  struct Entity {
+    Entity(const string &id) : id(id) {}
+    string id;
+    uint32 count = 0;
+  };
+
+  // Entity phrase with index and frequency. The count_and_flags field contains
+  // the count in the lower 29 bit. Bit 29 and 30 contain the case form, and
+  // bit 31 contains the reliable source flag.
+  struct EntityPhrase {
+    EntityPhrase() = default;
+    EntityPhrase(int index, uint32 count, uint32 form, bool reliable)
+        : index(index),
+          count_and_flags(count | (form << 29) | (reliable ? (1 << 31) : 0)) {}
+    uint32 index;
+    uint32 count_and_flags;
+
+    // Phrase frequency.
+    int count() const { return count_and_flags & ((1 << 29) - 1); }
+    void set_count(uint32 count) {
+      count_and_flags = (count_and_flags & ~((1 << 29) - 1)) | count;
+    }
+
+    // Alias reliability.
+    bool reliable() const { return count_and_flags & (1 << 31); }
+
+    // Phrase form.
+    int form() const { return (count_and_flags << 29) & 3; }
```

Comment:
shouldn't this be (count_and_flags >> 29) & 3 ?

Reply:
Oops. That's a bug. Good catch!